middleware/service/source/manager/state.cpp

```diff
@@ -320,69 +258,150 @@ namespace casual
+         }
 
-            instances.m_sequential.emplace_back( instance);
-            instance.add( *this);
+         
+         instance::concurrent::Property Service::property() const noexcept
+         {
+            if( ! has_sequential() && has_concurrent())
+               return { .hops = range::front( instances.concurrent()).hops};
+            
+            return {};
          }
 
-         void Service::add( state::instance::Concurrent& instance, state::service::instance::Concurrent::Property property)
+         state::service::id::type Services::insert( state::Service service)
          {
-            instances.add( instance, std::move( property));
+            // check if we've got the service before
+            if( auto id = lookup( service.information.name))
+               return id;
+
+            auto name = service.information.name;
+
+            auto id = m_services.insert( std::move( service));
+            m_lookup[ name] = id;
+            return id;
          }
- 
-         common::process::Handle Service::reserve_sequential(
-            const common::process::Handle& caller, 
-            const strong::correlation::id& correlation)
+
+         state::service::id::type Services::insert( state::Service service, const std::vector< std::string>& routes)
          {
-            if( auto found = algorithm::find_if( instances.m_sequential, []( auto& i){ return i.idle();}))
-               return found->reserve( this, caller, correlation);
+            auto id = [ &]()
+            {
+               // check if we've got the service by the origin name
+               if( auto id = lookup( service.information.name))
+                  return id;
 
-            return {};
+               // check if we've got any of the routes before
+               for( auto& name : routes)
+                  if( auto id = lookup( name))
+                     return id;
+
+               // we haven't got it, add it...
+               return m_services.insert( std::move( service));
+            }();
+
+            for( auto& name : routes)
+               m_lookup[ name] = id;
+
+            return id;
          }
 
-         common::process::Handle Service::reserve_concurrent( ipc_range_type preferred)
+         void Services::erase( state::service::id::type id)
          {
-            if( auto concurrent = instances.next_concurrent( preferred))
+            m_services.erase( id);
+            
+            algorithm::container::erase_if( m_lookup, [ this, id]( auto& pair)
             {
-               ++metric.remote;
-               return concurrent->process();
-            }            
-            return {};
+               if( pair.second != id)
+                  return false;
+
+               m_metrics.erase( pair.first);
+               return true;
+            });
          }
 
-         bool Service::is_discoverable() const noexcept
+         void Services::restore_routes( state::service::id::type id)
```

Comment:
So this erases all routes and restores the original name of the service? Would something like restore_origin_name or similar be more telling?

Reply:
I concur.